src/ir/can_derive_debug.rs

```diff
@@ -0,0 +1,269 @@
+            Some(&TypeKind::TemplateAlias(t, _)) |
+            Some(&TypeKind::Alias(t)) => {
+                if self.can_derive_debug.contains(&t) {
+                    self.can_derive_debug.insert(id);
+                    return true
+                }
+                false
+            },
+            Some(&TypeKind::Comp(ref info)) => {
+                if info.has_non_type_template_params() {
+                    if ty.unwrap().layout(self.ctx).map_or(false,
+                                                           |l| l.opaque().can_derive_debug(self.ctx, ())) {
+                        self.can_derive_debug.insert(id);
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+                if info.kind() == CompKind::Union {
+                    if self.ctx.options().unstable_rust {
+                        return false;
+                    }
+
+                    if ty.unwrap().layout(self.ctx).unwrap_or_else(Layout::zero)
+                        .opaque()
+                        .can_derive_debug(self.ctx, ()) {
+                            self.can_derive_debug.insert(id);
+                            return true;
+                        } else {
+                            return false;
+                        }
+                }
+                let consider_base = info.base_members()
+                    .iter()
+                    .all(|base| self.can_derive_debug.contains(&base.ty));
+                let consider_field = info.fields()
+                    .iter()
+                    .all(|f| {
+                        match f {
+                            &Field::DataMember(ref data) => self.can_derive_debug.contains(&data.ty()),
+                            &Field::Bitfields(ref bfu) => bfu.bitfields()
+                                .iter().all(|b| {
+                                    self.can_derive_debug.contains(&b.ty())
+                                })
+                        }
+                    });
+                if consider_base && consider_field {
+                    self.can_derive_debug.insert(id);
+                    return true;
+                }
+                false
+            },
+            Some(&TypeKind::Pointer(inner)) => {
+                let inner_type = self.ctx.resolve_type(inner);
+                if let TypeKind::Function(ref sig) =
+                    *inner_type.canonical_type (self.ctx).kind() {
+                        if sig.can_derive_debug(&self.ctx, ()) {
+                            self.can_derive_debug.insert(id);
+                            return true;
+                        } else {
+                            return false;
+                        }
+                    }
+                if self.can_derive_debug.contains(&inner) {
+                    self.can_derive_debug.insert(id);
+                    return true;
+                }
```

Comment:
There should be no check whether the pointed-to `inner` type is in the set or not: pointers can always derive `Debug` (its just their address) so the non-function pointer case should simply be this:

```rust
self.can_derive_debug.insert(id);
return true;
```